mesonbuild/msubprojects.py

```diff
@@ -0,0 +1,221 @@
+        # we should rebase it on top of wrap's branch.
+        if options.rebase:
+            try:
+                git(['fetch'], repo_dir)
+                git(['-c', 'rebase.autoStash=true', 'rebase', revision], repo_dir)
+            except subprocess.CalledProcessError as e:
+                out = e.output.decode().strip()
+                mlog.log('  -> Could not rebase', mlog.bold(repo_dir), 'please fix and try again.')
+                mlog.log(mlog.red(out))
+                mlog.log(mlog.red(str(e)))
+                return
+        else:
+            mlog.log('  -> Target revision is', mlog.bold(revision), 'but currently in branch is', mlog.bold(ret), '\n' +
+                     '     To rebase your branch on top of', mlog.bold(revision), 'use', mlog.bold('--rebase'), 'option.')
+            return
+
+    git(['submodule', 'update'], repo_dir)
+    git_show(repo_dir)
+
+def update_hg(wrap, repo_dir, options):
+    if not os.path.isdir(repo_dir):
+        mlog.log('  -> Not used.')
+        return
+    revno = wrap.get('revision')
+    if revno.lower() == 'tip':
+        # Failure to do pull is not a fatal error,
+        # because otherwise you can't develop without
+        # a working net connection.
+        subprocess.call(['hg', 'pull'], cwd=repo_dir)
+    else:
+        if subprocess.call(['hg', 'checkout', revno], cwd=repo_dir) != 0:
+            subprocess.check_call(['hg', 'pull'], cwd=repo_dir)
+            subprocess.check_call(['hg', 'checkout', revno], cwd=repo_dir)
+
+def update_svn(wrap, repo_dir, options):
+    if not os.path.isdir(repo_dir):
+        mlog.log('  -> Not used.')
+        return
+    revno = wrap.get('revision')
+    p, out = Popen_safe(['svn', 'info', '--show-item', 'revision', repo_dir])
+    current_revno = out
+    if current_revno == revno:
+        return
+    if revno.lower() == 'head':
+        # Failure to do pull is not a fatal error,
+        # because otherwise you can't develop without
+        # a working net connection.
+        subprocess.call(['svn', 'update'], cwd=repo_dir)
+    else:
+        subprocess.check_call(['svn', 'update', '-r', revno], cwd=repo_dir)
+
+def update(wrap, repo_dir, options):
+    mlog.log('Updating %s...' % wrap.name)
+    if wrap.type == 'file':
+        update_file(wrap, repo_dir, options)
+    elif wrap.type == 'git':
+        update_git(wrap, repo_dir, options)
+    elif wrap.type == 'hg':
+        update_hg(wrap, repo_dir, options)
+    elif wrap.type == 'svn':
+        update_svn(wrap, repo_dir, options)
+    else:
+        mlog.log('  -> Cannot update', wrap.type, 'subproject')
+
+def checkout(wrap, repo_dir, options):
+    if wrap.type != 'git' or not os.path.isdir(repo_dir):
+        return
+    branch_name = options.branch_name if options.branch_name else wrap.get('revision')
+    mlog.log('Checkout %s in %s...' % (branch_name, wrap.name))
+    try:
+        git(['checkout', '-B', branch_name, '--'], repo_dir)
```

Comment:
I think (from reading the docs) `-B` will clobber an existing branch. that does not sound like "checkout" to me.

Reply:
Oops you're right, that isn't the intention. I could just check if the branch exists before, but I wanted to profit from `git checkout foo` feature that create a branch if it exists in a remote. But I can't find a solution for that, so I think I'll have to add `meson subprojects checkout -b <branch>`... I'll think about that a bit more.

Reply:
Ok, updated my PR to add `-b` option to `meson subprojects checkout` that gets passed to `git checkout`.

I think the equivalent of what I wanted to do can still be done with 2 commands:
1. `meson subprojects checkout some_branch` will create branches that can be found in a remote, and fail for the subprojects that doesn't have it.
2. `meson subprojects checkout -b some_branch` will create the branch in the subprojects that failed in the first command.

I think that's good enough.